packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts

```diff
@@ -712,70 +726,53 @@ export class Connector implements IConnector {
         })();
     }
 
-    private isClosed() {
-        return new Promise(async (resolve) => {
-            let maxTries = 200;
-            let finish = false;
+    public async close() {
+        try {
 
-            if (!this._pid) {
-                resolve();
+            await this.lock();
 
-                return;
-            }
+            debug(`Pending tabs: ${this._tabs.length}`);
 
-            while (!finish) {
-                try {
-                    /*
-                     * We test if the process is still running or is a leftover:
-                     * https://nodejs.org/api/process.html#process_process_kill_pid_signal
-                     */
+            const { targetInfos } = await this._client.Target.getTargets!();
+            const targets = new Set();
+
+            targetInfos.forEach((target) => {
+                if (target.type === 'page') {
+                    targets.add(target.targetId);
+                }
+            });
 
-                    process.kill(this._pid, 0);
+            while (this._tabs.length > 0) {
+                const tab = this._tabs.pop();
 
-                    maxTries--;
+                try {
+                    if (targets.has(tab.id)) {
+                        targets.delete(tab.id);
 
-                    // Wait for 10 seconds to close the browser or continue.
-                    if (maxTries === 0) {
-                        finish = true;
-                    } else {
-                        await delay(50);
+                        await this._client.Target.closeTarget!({ targetId: tab.id });
                     }
                 } catch (e) {
-                    debug(`Process with ${this._pid} doesn't seem to be running`);
-                    finish = true;
+                    debug(`Couldn't close tab ${tab.id}`);
                 }
             }
 
-            resolve();
-        });
-    }
-
-    public async close() {
-        debug(`Pending tabs: ${this._tabs.length}`);
-
-        while (this._tabs.length > 0) {
-            const tab = this._tabs.pop();
-
             try {
-                await cdp.Close({ id: tab.id, port: (this._client as any).port }); // eslint-disable-line new-cap
+                (this._client as any).close();
+
+                if (targets.size === 0 && this._pid) {
```

Comment:
For future reference we should put some version of your comment from the PR description either here or at the top of `close`:
> On macOS Chrome does not fully exit even if all the tabs are closed.
This causes a timeout when running the tests with nyc of about 2
minutes (or if the browser is manually closed). This addresses this
issue.
Because during testing multiple workers are using the same instance of
Chrome, one could try to close the browser while another is opening a
new tab and thus causing communication issues. The solution is to use
the same lock to launch the browser and to close it with a twist: The
close method gets all the possible Targets of type page (tabs),
close the ones related to the current session and see if there are any
other left. If there are not, that means it is safe to completely close
the browser by sending a SIGKILL (no other methods works because the
debugging protocol client cannot communicate in this state and
gracefully close the browser).

Reply:
Sounds good. I might have to tweak it after the castatrophic failure in CI. Maybe ava is doing something clever underneath when it detects that is CI that doesn't work nice with my changes (less workers?). I'm going to try to run this PR locally on Windows and see if I have to enable the kill signal only outside CI and/or macs

Reply:
Looks like this only repros in CI only. Tests are running fine on my Windows machine too 😒

Reply:
So the plot thickens...

If you set the environemt variable `SYSTEM_TEAMFOUNDATIONCOLLECTIONURI` to something so `is-ci` returns `true`, tests start blocking. At the beginning I thought it was something related to `chrome` and my changes but if I disable the `chrome` connector on `hint-amp-validator` it looks like the problem is after finishing the first `jsdom` test 🤯

```
Debugger attached.

  - [chrome] Valid AMP HTML passes
  - [chrome] Resource is not an HTML document
  - [chrome] Invalid AMP HTML fails
  - [chrome] Error downloading HTML doesn't fail
  √ [jsdom] Resource is not an HTML document (2s)
```

I have no idea what is going on